test/framework/src/main/java/org/elasticsearch/test/InternalTestCluster.java

```diff
@@ -396,6 +396,14 @@ private Settings getRandomNodeSettings(long seed) {
             builder.put(MappingUpdatedAction.INDICES_MAPPING_DYNAMIC_TIMEOUT_SETTING.getKey(), new TimeValue(RandomNumbers.randomIntBetween(random, 10, 30), TimeUnit.SECONDS));
         }
 
+        if (random.nextBoolean()) {
+            builder.put(HierarchyCircuitBreakerService.USE_REAL_MEMORY_USAGE_SETTING.getKey(), true);
+            // allow full allocation in tests to avoid breaking
```

Comment:
We do. I have the impression that this is mainly caused by two facts: 

1. We run multiple integration tests in the same JVM process so garbage will add up.
2. We do not set a garbage collection algorithm in the integration tests (except for our g1 builds via `-Dtests.jvm.argline` in the Jenkins configuration). This means we let the JVM choose the algorithm. For the Java 10 builds this is `G1`, for Java 8 it is `ParallelGC`. From my tests I had the impression that especially G1 had trouble keeping up with allocation in our tests (as G1 cleans up concurrently) and I experienced circuit breaker issues basically all the time. 

 I am not sure how we should move forward here in general. G1 is not really optimal for such small heaps but on the other hand we should also run our tests with G1. I already added this to our group's agenda to discuss. Finally, the 95% are heap usage + amount to be reserved.

Reply:
> Finally, the 95% are heap usage + amount to be reserved.

I suppose in that case it would force a real GC if the heap were at 95% and then a new amount was reserved using `new`

Reply:
Actually the GC should get triggered even earlier. For CMS we have configured `-XX:CMSInitiatingOccupancyFraction=75`. But as the garbage collector is running concurrently with the application, it can still allocate while the GC is active.